[Detection Engine][Exceptions] - Fix exception item update route

yctercero opened the pull request:
## Summary

Addresses https://github.com/elastic/kibana/issues/159230. Fixes a bug in exception item update route. When updating an exception item and not passing in `id`, `id` [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/lists/server/services/exception_lists/update_overwrite_exception_list_item.ts#L76) is undefined and because it does not see an SO to overwrite, it creates it.

Wrote test to confirm that this behavior is occurring: 
<img width="819" alt="Screenshot 2023-06-07 at 8 13 09 AM" src="https://github.com/elastic/kibana/assets/10927944/731392f5-ff85-475e-97e8-033ee4b0c510">

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

<!--ONMERGE {"backportTargets":["8.7","8.8"]} ONMERGE-->

yctercero committed:
fix exception item update route

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

yctercero committed:
pr feedback updates

rylnd commented on x-pack/test/lists_api_integration/security_and_spaces/tests/update_exception_list_items.ts:
```diff
@@ -31,6 +31,60 @@ export default ({ getService }: FtrProviderContext) => {
         await deleteAllExceptions(supertest, log);
       });
 
+      // SDH came up where this route was creating a new item when updating
+      // by item_id, ensuring that's no longer the case
+      it('should not create a new item on update', async () => {
```
I think we could do away with the comment and put this context into the test description. I deleted the mention of an SDH because IMO the fact that it's a regression test isn't particularly important; deleting a test is an exception for which one had better have a good reason!

```suggestion
      it('should not create a new item when updating by item_id', async () => {
```

rylnd reviewed (approved):
LGTM once the `id` arg is respected and we're 🟢 .

rylnd committed:
Add another regression test to cover the inverse bug

Similar to how specifying item_id when updating causes a new id to be
generated, specifying an id causes the item_id to be deleted if
unspecified.

This commit adds a test to cover this case, along with updating the
existing regression to have more assertions and have more accurate
variable names.

rylnd committed:
Nest our tests to reflect their role as regression tests

rylnd committed:
Add broader regression test for what is preserved on update

This test attempts to see exactly what's going on with this endpoint,
currently:

1. Create a fully custom, fully specified exception item
    * WIP; see TODOs in code
2. Update it with only the required fields
3. See what changed after the update

yctercero committed:
Merge branch 'exceptions_update_route' of github.com:yctercero/kibana into exceptions_update_route

yctercero committed:
Merge branch 'main' of github.com:elastic/kibana into exceptions_update_route

yctercero committed:
fix item_id issue

rylnd committed:
Update 'catch all' regression test

* More descriptive test/comments
* Remove server-generated noise from our assertions
* Add comments to the item

This test is currently failing due to the fact that we drop tags if they
aren't specified. Will see if this behavior is consistent with the old
implementation.

rylnd committed:
Remove tags from our 'catch all' regression test

If we specify custom tags here, the test will fail because tags are
overwritten if not specified on update.

However, this has always been the behavior of this endpoint (it at least
preceded the changes causing the id-related bugs), so we're leaving it
for now.

yctercero commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into exceptions_update_route

rylnd committed:
Merge branch 'main' into exceptions_update_route

rylnd commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into exceptions_update_route

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.7|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/159418">](https://github.com/elastic/kibana/pull/159418)|
|✅|8.8|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/159420">](https://github.com/elastic/kibana/pull/159420)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.
